### PR TITLE
Add pipeline draft CRUD endpoints for RDI instances

### PR DESCRIPTION
**Pull request by ArtemHoruzhenko**
# What

Third PR in the rdi-ui integration chain (stacked on #6465, which is stacked on #6464).

Adds a `pipeline_draft` table and CRUD endpoints under `rdi/:id/pipeline-drafts` (list/get/create/update/delete), so RDI pipeline work-in-progress can be persisted per instance ahead of the real `rdi-ui` component being wired in.

Key points:
- **Orphan removal**: FK `pipeline_draft.rdiInstanceId -> rdi.id` with `ON DELETE CASCADE` at the DB level, mirroring the `query-library` module's existing FK-to-parent pattern. No manual cleanup code needed when an RDI instance is deleted.
- **Encryption**: `data` is encrypted at rest via the existing `ModelEncryptor`, same mechanism as `RdiEntity.password` and `QueryLibraryEntity`'s fields.
- **JSON, not string**: the entity column has to stay a `text`/string column (encryption only operates on strings), but it's decorated with the existing `@DataAsJsonString()` decorator so every layer above the entity (API model, DTOs, controllers) works with a real JSON object, not a string — same mechanism already used by `AiQueryMessageEntity.steps`. Structure of the JSON itself is intentionally not validated/typed.
- Multiple drafts per RDI instance are supported (not a 1:1 relationship), with a full update endpoint (not just create/delete).
- Followed `RdiPipelineController`'s existing `rdi/:id/...` nesting convention and `query-library`'s repository/service/controller layering as templates.

# Testing

- `npm run lint:api` and `npm run type-check --prefix redisinsight/api` (0 new errors) pass.
- 3 new spec files (26 tests) covering the repository (incl. cascade-relevant FK setup, encryption interaction, and a dedicated test proving the entity round-trips `data` as JSON string internally while exposing a real object externally), service, and controller.
- Full `rdi` module test suite (286 tests) still passes.

---

No ticket yet.

---

> [!NOTE]
> **Medium Risk**
> New encrypted persistence and CRUD surface for RDI draft data; repository edge cases around decrypt failures affect data visibility and updates, but patterns match existing RDI/query-library modules.
> 
> **Overview**
> Adds **persisted RDI pipeline drafts** so in-progress pipeline JSON can be saved per instance before the UI is wired up.
> 
> A new `pipeline_draft` table (migration + TypeORM entity) stores encrypted draft `data` keyed by `rdiInstanceId`, with **ON DELETE CASCADE** when an RDI instance is removed. The API exposes full CRUD at `rdi/:id/pipeline-drafts`, using the existing `RequestRdiClientMetadata` nesting and validation (create whitelists body fields; update rejects explicit `null` for `data`). Create verifies the parent RDI exists via `RdiRepository.get(..., true)` without decrypting credentials.
> 
> Persistence goes through `LocalPipelineDraftRepository`, which encrypts `data` with `ModelEncryptor`, exposes JSON objects via `@DataAsJsonString()`, skips undecryptable rows on list, and avoids overwriting ciphertext on failed decrypt during update. `RdiModule` and ORM config register the new stack; controller/service/repository specs cover the behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e20ca94fb77a02ae7a4a37eaec4ace86ef0f0ab7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+  async create(
+    sessionMetadata: SessionMetadata,
+    rdiInstanceId: string,
+    data: Partial<PipelineDraft>,
+  ): Promise<PipelineDraft> {
+    this.logger.debug('Creating pipeline draft', sessionMetadata);
+
+    const entity = plainToInstance(PipelineDraftEntity, {
+      ...data,
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:37]: fixed in [c4db648](https://github.com/redis/RedisInsight/pull/6467/commits/c4db648d2f)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      );
+      throw new NotFoundException(`Pipeline draft with id ${id} was not found`);
+    }
+
+    const decrypted = await this.modelEncryptor.decryptEntity(existing, true);
+    const updateData = omitBy(data, isUndefined);
+
+    const merged = plainToInstance(PipelineDraftEntity, {
+      ...decrypted,
+      ...updateData,
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts]: fixed in [c4db648](https://github.com/redis/RedisInsight/pull/6467/commits/c4db648d2f)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+      id,
+      rdiInstanceId,
+    });
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:136]: fixed in [c4db648](https://github.com/redis/RedisInsight/pull/6467/commits/c4db648d2f)

File: redisinsight/api/src/modules/rdi/dto/update.pipeline-draft.dto.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { PartialType } from '@nestjs/swagger';
+import { CreatePipelineDraftDto } from 'src/modules/rdi/dto/create.pipeline-draft.dto';
+
+export class UpdatePipelineDraftDto extends PartialType(
+  CreatePipelineDraftDto,
+) {}
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/dto/update.pipeline-draft.dto.ts]: fixed in [0d988a4](https://github.com/redis/RedisInsight/pull/6467/commits/0d988a4dfe)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    this.logger.debug('Creating pipeline draft', sessionMetadata);
+
+    const entity = plainToInstance(PipelineDraftEntity, {
+      ...data,
+      rdiInstanceId,
+    });
+
+    const saved = await this.repository.save(
+      await this.modelEncryptor.encryptEntity(entity),
+    );
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:43]: fixed in [0d988a4](https://github.com/redis/RedisInsight/pull/6467/commits/0d988a4dfe)

File: redisinsight/api/src/modules/rdi/entities/pipeline-draft.entity.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+@Entity('pipeline_draft')
+export class PipelineDraftEntity {
+  @Expose()
+  @PrimaryGeneratedColumn('uuid')
+  id: string;
+
+  @Expose()
+  @Column({ nullable: false })
+  rdiInstanceId: string;
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/entities/pipeline-draft.entity.ts:24]: fixed in [0d988a4](https://github.com/redis/RedisInsight/pull/6467/commits/0d988a4dfe)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+    // JSON.stringify an already-stringified value
+    const merged = plainToInstance(PipelineDraftEntity, {
+      ...restDecrypted,
+      ...restUpdateData,
+      id,
+      rdiInstanceId,
+    });
+    merged.data = isUndefined(updatedData)
+      ? decryptedData
+      : JSON.stringify(updatedData);
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:139]: fixed in [0d988a4](https://github.com/redis/RedisInsight/pull/6467/commits/0d988a4dfe)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        try {
+          return await this.modelEncryptor.decryptEntity(entity);
+        } catch (e) {
+          return null;
+        }
+      }),
+    );
+
+    return filter(decryptedEntities, (entity) => !isNull(entity)).map(
+      (entity) => classToClass(PipelineDraft, entity),
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts]: fixed in [39afac0](https://github.com/redis/RedisInsight/pull/6467/commits/39afac0654)

File: redisinsight/api/src/modules/rdi/pipeline-draft.service.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  ) {}
+
+  async create(
+    sessionMetadata: SessionMetadata,
+    rdiInstanceId: string,
+    dto: CreatePipelineDraftDto,
+  ): Promise<PipelineDraft> {
+    this.logger.debug('Creating pipeline draft', sessionMetadata);
+
+    const rdi = await this.rdiRepository.get(rdiInstanceId);
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/pipeline-draft.service.ts]: fixed in [39afac0](https://github.com/redis/RedisInsight/pull/6467/commits/39afac0654)

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        sessionMetadata,
+      );
+      throw new NotFoundException(`Pipeline draft with id ${id} was not found`);
+    }
+
+    this.logger.debug('Succeed to get pipeline draft', sessionMetadata);
+
+    return classToClass(
+      PipelineDraft,
+      await this.modelEncryptor.decryptEntity(entity, true),
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:101]: it is the same behavior as everywhere else. if we are not able to decrypt - it's better to start from the scratch

**Review by pawelangelow — APPROVED** [review]

File: redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      id,
+      rdiInstanceId,
+    });
+    merged.data = isUndefined(updatedData)
+      ? decryptedData
+      : JSON.stringify(updatedData);
+
+    const saved = await this.repository.save(
+      await this.modelEncryptor.encryptEntity(merged),
+    );
```
[the remark this answers is not in the document]
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/repository/local.pipeline-draft.repository.ts:143]: This read-modify-write-via-save() shape (and its resurrection race) is the existing convention across this codebase - the same pattern exists in local.database.repository.ts, local.cloud-capi-key.repository.ts, and local.rdi.repository.ts, none of which guard against it either. Treating this as a pre-existing, repo-wide design tradeoff rather than something to special-case here; a proper fix would mean auditing all of them together, not just this one.
